Localize the client/browser UI with inlang paraglide-js

IDisposable opened the pull request:
### Localization

The browser/client frontend uses the [paraglide-js](https://inlang.com/m/gerre34r/library-inlang-paraglideJs) plug-in from the [inlang.com](https://inlang.com/) project to allow compile-time validated localization of all user-facing UI strings in the browser/client UI. This includes `title`, `text`, `name`, `description`, `placeholder`, `label`, `aria-label`, _message attributes_ (such as `confirmText`, `unit`, `badge`, `tag`, or `flag`), HTML _element text_ (such as `<h?>`, `<span>`, or `<p>` elements), _notifications messages_, and option _label_ strings, etc.

We **do not** translate the console log messages, CSS class names, theme names, nor the various _value_ strings (e.g. for value/label pair options), nor URL routes.

The localizations are stored in _.json_ files in the `ui/localizations/messages` directory, with one language-per-file using the [ISO 3166-1 alpha-2 country code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) (e.g. en for English, de for German, etc.)

#### m-function-matcher

The translations are extracted into language files (e.g. _en.json_ for English) and then paraglide-js compiles them into helpers for use with the [m-function-matcher](https://inlang.com/m/632iow21/plugin-inlang-mFunctionMatcher). An example:

```tsx
<SettingsPageHeader
   title={m.extensions_atx_power_control()}
   description={m.extensions_atx_power_control_description()}
/>
```

#### shakespere plug-in

If you enable the [Sherlock](https://inlang.com/m/r7kp499g/app-inlang-ideExtension) plug-in, the localized text "tooltip" is shown in the VSCode editor after any localized text in the language you've selected for preview. In this image, it's the blue text at the end of the line :

![Showing the translation preview](https://github.com/user-attachments/assets/f6d6dae6-919f-4319-b7bf-500cb1fd458d)

#### Process

##### Localizing a UI

1. Locate a string that is visible to the end user on the client/browser
2. Assign that string a "key" that reflects the logical meaning of the string in snake-case (look at existing localizations for examples), for example if there's a string `This is a test` on the _thing edit page_ it would be "thing_edit_this_is_a_test"

   ```json
   "thing_edit_this_is_a_test": "This is a test",
   ```

3. Add the key and string to the _en.json_ like this:

   - **Note** if the string has replacement parameters (line a user-entered name), the syntax for the localized string has `{ }` around the replacement token (e.g. _This is your name: {name}_). An complex example:

   ```react
   {m.mount_button_showing_results({
      from: indexOfFirstFile + 1,
      to: Math.min(indexOfLastFile, onStorageFiles.length),
      total: onStorageFiles.length
   })}
   ```

4. Save the _en.json_ file and execute `npm run i18n` to resort the language files, validate the translations, and create the m-functions
5. Edit the _.tsx_ file and replace the string with the calls to the new m-function which will be the key-string you chose in snake-case. For example `This is a test` in _thing edit page_ turns into `m.thing_edit_this_is_a_test()`
   - **Note** if the string has a replacement token, supply that to the m-function, for example for the literal `I will call you {name}`, use `m.profile_i_will_call_you({ name: edit.value })`
6. When all your strings are extracted, run `npm run i18n:machine-translate` to get a first-stab at the translations for the other supported languages. Make sure you use an LLM (you can use [aifiesta](https://chat.aifiesta.ai/chat/) to use multiple LLMs) or a [translator](https://translate.google.com) of some form to back-translate each **new** machine-generation in each _language_ to ensure those terms translate reasonably.

### Adding a new language

1. Get the [ISO 3166-1 alpha-2 country code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) (for example AT for Austria)
2. Create a new file in the _ui/localization/messages_ directory (example _at.json_)
3. Add the new country code to the _ui/localizations/settings.json_ file in both the `"locales"` and the `"languageTags"` section (inlang and Sherlock aren't exactly current to each other, so we need it in both places).
4. That file also declares the baseLocale/sourceLanguageTag which is `"en"` because this project started out in English. Do NOT change that.
5. Run `npm run i18n:machine-translate` to do an initial pass at localizing all existing messages to the new language.
   - **Note** you will get an error _DB has been closed_, ignore that message, we're not using a database.
   - **Note** you likely will get errors while running this command due to rate limits and such (it uses anonymous Google Translate). Just keep running the command over and over... it'll translate a bunch each time until it says _Machine translate complete_

### Other notes

- Run `npm run i18n:validate` to ensure that language files and settings are well-formed.
- Run `npm run i18n:find-excess` to look for extra keys in other language files that have been deleted from the master-list in _en.json_.
- Run `npm run i18n:find-dupes` to look for multiple keys in _en.json_ that have the same translated value (this is normal)
- Run `npm run i18n:find-unused` to look for keys in _en.json_ that are not referenced in the UI anywhere.
  - **Note** there are a few that are not currently used, only concern yourself with ones you obsoleted.
- Run `npm run i18n:audit` to do all the above checks.
- Using [inlang CLI](https://inlang.com/m/2qj2w8pu/app-inlang-cli) to support the npm commands.
- You can install the [Sherlock VS Code extension](https://marketplace.visualstudio.com/items?itemName=inlang.vs-code-extension) in your devcontainer.

IDisposable commented on ui/src/components/extensions/SerialConsole.tsx:
```diff
@@ -66,18 +63,18 @@ export function SerialConsole() {
+              text={m.serial_console_open_console()}
               onClick={() => {
-                setTerminalType("serial");
                 console.log("Opening serial console with settings: ", settings);
+                setTerminalType("serial");
```
Nope, that was intentional

IDisposable committed:
Add inlang/paraglide-js localization

Remove the temporary directory after extracting buildkit
Localize the extension popovers.
Update package and fix tsconfig.json
Expand development directory guide
Move messages under localization
Popovers and sidebar
Update Chinese translations
Accidentally lost the changes that @ym provided, brought them back
File formatting pass
Localized all components, hooks, providers, hooks
Localize all pages except Settings
Bump packages
Settings Access page
Settings local auth page
Fix ref lint warning
Settings Advanced page
Fix UI lint warnings there were a bunch of ref and useEffect violations.
Settings appearance page
Settings general pages
Settings hardware page
Settings keyboard page
Settings macros pages
Settings mouse page
Settings page
Settings video page
Settings network page
Fix compilation issues
Ran machine translate
Use getLocale for date, relative time, and money formatting
Fix eslint
Delete unused messages
Added setting to choose locale
Merged in dev hotfix
Fix update status rendering
Add note to do back-translation
Improve developer guidance
Clean up some localization issues and extract UpdatingStatusCard
Fix copy-pasta errors
Packages update
Fix silly error in progress messages.

IDisposable commented on ui/localization/jetKVM.UI.inlang/settings.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "$schema": "https://inlang.com/schema/project-settings",
+  "baseLocale": "en",
+  "sourceLanguageTag": "en",
```
This duplication is because the CLI is a bit behind the extension Sherlock, so both are needed. PR being submitted to the inlang project

IDisposable commented on ui/src/components/MetricsChart.tsx:
```diff
@@ -51,6 +51,7 @@ export default function MetricsChart({
           tickLine={{ stroke: "rgba(30, 41, 59, 0.3)" }}
           tickFormatter={date => {
+            // TODO use locale from user settings for date formatting
             return new Date(date * 1000).toLocaleString("en-US", {
```
We need to think through swapping uses of toLocaleString to appropriate Intl as appropriate. At minimum need to get the current local from the browser.

IDisposable replied:
Fixed

IDisposable commented on ui/src/components/MetricsChart.tsx:
```diff
@@ -51,6 +51,7 @@ export default function MetricsChart({
           tickLine={{ stroke: "rgba(30, 41, 59, 0.3)" }}
           tickFormatter={date => {
+            // TODO use locale from user settings for date formatting
             return new Date(date * 1000).toLocaleString("en-US", {
```
Actually, we can get the effective locale from 

`import { getLocale } from '@localizations/runtime.js';`

IDisposable replied:
Done.

IDisposable commented on ui/src/components/MacroStepCard.tsx:
```diff
@@ -25,6 +26,7 @@ const groupedModifiers: Record<string, typeof modifierOptions> = {
   Meta: modifierOptions.filter(mod => mod.value.startsWith('Meta')),
 };
 
+// not going to localize these since they're short time intervals
 const basePresetDelays = [
```
Yes, that's what a conscious decision is for...

IDisposable commented:
I think this is ready for testing, the Update page is still borked, but will fix that ASAP. You can use a cookie-editor to switch languages until someone weighs in on how we want to present the UI language switcher on the client

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "access_no_device_id": "Inget enhets-ID tillgängligt",
+    "access_private_key_description": "Av säkerhetsskäl kommer den inte att visas efter att den har sparats.",
+    "access_private_key_label": "Privat nyckel",
+    "access_provider_custom": "Beställnings",
```
"access_provider_custom": "Anpassad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "access_tls_certificate_description": "Klistra in ditt TLS-certifikat nedan. För certifikatkedjor, inkludera hela kedjan (löv-, mellan- och rotcertifikat).",
+    "access_tls_certificate_title": "TLS-certifikat",
+    "access_tls_custom": "Beställnings",
+    "access_tls_disabled": "Funktionshindrad",
```
"access_tls_custom": "Anpassad",
    "access_tls_disabled": "Inaktiverad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "back": "Tillbaka",
+    "back_to_devices": "Tillbaka till Enheter",
+    "cancel": "Avboka",
+    "close": "Nära",
```
"cancel": "Avbryt",
    "close": "Stäng",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "local_auth_create_new_password_label": "Nytt lösenord",
+    "local_auth_create_new_password_placeholder": "Ange ett starkt lösenord",
+    "local_auth_create_not_now_button": "Inte nu",
+    "local_auth_create_secure_button": "Säker enhet",
```
"local_auth_create_secure_button": "Säkra enheten",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "locale_en": "Engelska",
+    "locale_es": "Spanska",
+    "locale_fr": "Franska",
+    "locale_it": "italiensk",
```
"locale_auto": "Auto",
    "locale_change_success": "Språket har ändrats till {locale}",
    "locale_da": "Danska",
    "locale_de": "Tyska",
    "locale_en": "Engelska",
    "locale_es": "Spanska",
    "locale_fr": "Franska",
    "locale_it": "Italienska",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "mouse_jiggler_config_updated": "Jiggler-konfigurationen har uppdaterats",
+    "mouse_jiggler_custom": "Beställnings",
+    "mouse_jiggler_description": "Simulera rörelsen hos en datormus",
+    "mouse_jiggler_disabled": "Funktionshindrad",
```
"mouse_jiggler_custom": "Anpassad",
    "mouse_jiggler_description": "Simulera rörelsen hos en datormus",
    "mouse_jiggler_disabled": "Inaktiverad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "network_dhcp_lease_renew_confirm_new_b": "du kan behöva återansluta med den nya adressen",
+    "network_dhcp_lease_renew_failed": "Misslyckades med att förnya leasingavtalet: {error}",
+    "network_dhcp_lease_renew_success": "DHCP-lease förnyad",
+    "network_domain_custom": "Beställnings",
```
"network_domain_custom": "Anpassad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "network_ipv6_information": "IPv6-information",
+    "network_ipv6_mode_description": "Konfigurera IPv6-läget",
+    "network_ipv6_mode_dhcpv6": "DHCPv6",
+    "network_ipv6_mode_disabled": "Funktionshindrad",
```
"network_ipv6_mode_disabled": "Inaktiverad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "network_ll_dp_all": "Alla",
+    "network_ll_dp_basic": "Grundläggande",
+    "network_ll_dp_description": "Kontrollera vilka TLV:er som ska skickas via Link Layer Discovery Protocol",
+    "network_ll_dp_disabled": "Funktionshindrad",
```
"network_ll_dp_disabled": "Inaktiverad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "network_mac_address_title": "MAC-adress",
+    "network_mdns_auto": "Bil",
+    "network_mdns_description": "Kontroll av mDNS (multicast DNS) driftläge",
+    "network_mdns_disabled": "Funktionshindrad",
```
"network_mdns_auto": "Auto",
    "network_mdns_description": "Kontroll av mDNS (multicast DNS) driftläge",
    "network_mdns_disabled": "Inaktiverad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "updates_failed_get_device_version": "Misslyckades med att hämta enhetsversionen: {error}",
+    "updating_leave_device_on": "Stäng inte av din enhet…",
+    "usb": "USB",
+    "usb_config_custom": "Beställnings",
```
"usb_config_custom": "Anpassad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "usb_config_vendor_id_placeholder": "Ange leverantörs-ID",
+    "usb_device_classes_description": "USB-enhetsklasser i den sammansatta enheten",
+    "usb_device_classes_title": "Klasser",
+    "usb_device_custom": "Beställnings",
```
"usb_device_custom": "Anpassad",

adamshiervani commented on ui/localization/messages/sv.json:
```diff
@@ -0,0 +1,876 @@
+    "video_description": "Konfigurera skärminställningar och EDID för optimal kompatibilitet",
+    "video_edid_acer_b246wl": "Acer B246WL, 1920x1200",
+    "video_edid_asus_pa248qv": "ASUS PA248QV, 1920x1200",
+    "video_edid_custom": "Beställnings",
```
"video_edid_custom": "Anpassad",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "access_no_device_id": "Intet enheds-ID tilgængeligt",
+    "access_private_key_description": "Af sikkerhedsmæssige årsager vil den ikke blive vist efter lagring.",
+    "access_private_key_label": "Privat nøgle",
+    "access_provider_custom": "Skik",
```
"access_provider_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "access_tls_certificate_description": "Indsæt dit TLS-certifikat nedenfor. For certifikatkæder skal du inkludere hele kæden (blad-, mellem- og rodcertifikater).",
+    "access_tls_certificate_title": "TLS-certifikat",
+    "access_tls_custom": "Skik",
+    "access_tls_disabled": "Handicappet",
```
"access_tls_custom": "Tilpasset",
    "access_tls_disabled": "Deaktiveret",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "atx_power_control_get_state_error": "Kunne ikke hente ATX-strømtilstand: {error}",
+    "atx_power_control_hdd_led": "HDD-LED",
+    "atx_power_control_long_power_button": "Langt tryk",
+    "atx_power_control_power_button": "Magt",
```
"atx_power_control_power_button": "Strøm",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "back": "Tilbage",
+    "back_to_devices": "Tilbage til Enheder",
+    "cancel": "Ophæve",
+    "close": "Tæt",
```
"cancel": "Annuller",
    "close": "Luk",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "dc_power_control_current": "Strøm",
+    "dc_power_control_current_unit": "EN",
+    "dc_power_control_get_state_error": "Kunne ikke hente DC-strømtilstand: {error}",
+    "dc_power_control_power": "Magt",
```
"dc_power_control_power": "Strøm",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "local_auth_create_new_password_label": "Ny adgangskode",
+    "local_auth_create_new_password_placeholder": "Indtast en stærk adgangskode",
+    "local_auth_create_not_now_button": "Ikke nu",
+    "local_auth_create_secure_button": "Sikker enhed",
```
"local_auth_create_secure_button": "Sikr enheden",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "local_auth_success_password_updated_description": "Du har ændret din adgangskode til beskyttelse af din lokale enhed. Husk din nye adgangskode til senere brug.",
+    "local_auth_success_password_updated_title": "Adgangskode opdateret",
+    "local_auth_update_password_button": "Opdater adgangskode",
+    "locale_auto": "Bil",
```
"locale_auto": "Auto",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "mouse_jiggler_config_updated": "Jiggler-konfigurationen er blevet opdateret",
+    "mouse_jiggler_custom": "Skik",
+    "mouse_jiggler_description": "Simuler bevægelsen af en computermus",
+    "mouse_jiggler_disabled": "Handicappet",
```
"mouse_jiggler_custom": "Tilpasset",
    "mouse_jiggler_description": "Simuler bevægelsen af en computermus",
    "mouse_jiggler_disabled": "Deaktiveret",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "network_dhcp_lease_renew_confirm_new_b": "du skal muligvis genoprette forbindelsen ved hjælp af den nye adresse",
+    "network_dhcp_lease_renew_failed": "Kunne ikke forny leasing: {error}",
+    "network_dhcp_lease_renew_success": "DHCP-lease fornyet",
+    "network_domain_custom": "Skik",
```
"network_domain_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "network_ipv6_information": "IPv6-oplysninger",
+    "network_ipv6_mode_description": "Konfigurer IPv6-tilstanden",
+    "network_ipv6_mode_dhcpv6": "DHCPv6",
+    "network_ipv6_mode_disabled": "Handicappet",
```
"network_ipv6_mode_disabled": "Deaktiveret",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "network_ll_dp_all": "Alle",
+    "network_ll_dp_basic": "Grundlæggende",
+    "network_ll_dp_description": "Styr hvilke TLV'er der sendes via Link Layer Discovery Protocol",
+    "network_ll_dp_disabled": "Handicappet",
```
"network_ll_dp_disabled": "Deaktiveret",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "network_mac_address_title": "MAC-adresse",
+    "network_mdns_auto": "Bil",
+    "network_mdns_description": "Styr mDNS (multicast DNS) driftstilstand",
+    "network_mdns_disabled": "Handicappet",
```
"network_mdns_auto": "Auto",
    "network_mdns_description": "Styr mDNS (multicast DNS) driftstilstand",
    "network_mdns_disabled": "Deaktiveret",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "updates_failed_get_device_version": "Kunne ikke hente enhedsversion: {error}",
+    "updating_leave_device_on": "Sluk venligst ikke din enhed…",
+    "usb": "USB",
+    "usb_config_custom": "Skik",
```
"usb_config_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "usb_config_vendor_id_placeholder": "Indtast leverandør-ID",
+    "usb_device_classes_description": "USB-enhedsklasser i den sammensatte enhed",
+    "usb_device_classes_title": "Klasser",
+    "usb_device_custom": "Skik",
```
"usb_device_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/da.json:
```diff
@@ -0,0 +1,876 @@
+    "video_description": "Konfigurer skærmindstillinger og EDID for optimal kompatibilitet",
+    "video_edid_acer_b246wl": "Acer B246WL, 1920x1200",
+    "video_edid_asus_pa248qv": "ASUS PA248QV, 1920x1200",
+    "video_edid_custom": "Skik",
```
"video_edid_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "access_no_device_id": "Ingen enhets-ID tilgjengelig",
+    "access_private_key_description": "Av sikkerhetshensyn vil den ikke vises etter lagring.",
+    "access_private_key_label": "Privat nøkkel",
+    "access_provider_custom": "Skikk",
```
"access_provider_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "access_tls_certificate_description": "Lim inn TLS-sertifikatet ditt nedenfor. For sertifikatkjeder, inkluder hele kjeden (blad-, mellom- og rotsertifikater).",
+    "access_tls_certificate_title": "TLS-sertifikat",
+    "access_tls_custom": "Skikk",
+    "access_tls_disabled": "Funksjonshemmet",
```
"access_tls_custom": "Tilpasset",
    "access_tls_disabled": "Deaktivert",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "atx_power_control_get_state_error": "Klarte ikke å hente ATX-strømstatus: {error}",
+    "atx_power_control_hdd_led": "HDD-LED",
+    "atx_power_control_long_power_button": "Langt trykk",
+    "atx_power_control_power_button": "Makt",
```
"atx_power_control_power_button": "Strøm",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "back": "Tilbake",
+    "back_to_devices": "Tilbake til Enheter",
+    "cancel": "Kansellere",
+    "close": "Lukke",
```
"cancel": "Avbryt",
    "close": "Lukk",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "dc_power_control_current": "Nåværende",
+    "dc_power_control_current_unit": "EN",
+    "dc_power_control_get_state_error": "Klarte ikke å hente likestrømsstatus: {error}",
+    "dc_power_control_power": "Makt",
```
"dc_power_control_power": "Strøm",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "local_auth_create_new_password_label": "Nytt passord",
+    "local_auth_create_new_password_placeholder": "Skriv inn et sterkt passord",
+    "local_auth_create_not_now_button": "Ikke nå",
+    "local_auth_create_secure_button": "Sikker enhet",
```
"local_auth_create_secure_button": "Sikre enheten",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "local_auth_success_password_updated_description": "Du har endret passordet for beskyttelse av den lokale enheten. Husk det nye passordet for fremtidig tilgang.",
+    "local_auth_success_password_updated_title": "Passord oppdatert",
+    "local_auth_update_password_button": "Oppdater passord",
+    "locale_auto": "Bil",
```
"locale_auto": "Auto",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "mouse_jiggler_config_updated": "Jiggler-konfigurasjonen er oppdatert",
+    "mouse_jiggler_custom": "Skikk",
+    "mouse_jiggler_description": "Simuler bevegelsen til en datamus",
+    "mouse_jiggler_disabled": "Funksjonshemmet",
```
"mouse_jiggler_custom": "Tilpasset",
    "mouse_jiggler_description": "Simuler bevegelsen til en datamus",
    "mouse_jiggler_disabled": "Deaktivert",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "network_dhcp_lease_renew_confirm_new_b": "du må kanskje koble til på nytt med den nye adressen",
+    "network_dhcp_lease_renew_failed": "Kunne ikke fornye leieavtalen: {error}",
+    "network_dhcp_lease_renew_success": "DHCP-leieavtale fornyet",
+    "network_domain_custom": "Skikk",
```
"network_domain_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "network_ipv6_information": "IPv6-informasjon",
+    "network_ipv6_mode_description": "Konfigurer IPv6-modusen",
+    "network_ipv6_mode_dhcpv6": "DHCPv6",
+    "network_ipv6_mode_disabled": "Funksjonshemmet",
```
"network_ipv6_mode_disabled": "Deaktivert",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "network_ll_dp_all": "Alle",
+    "network_ll_dp_basic": "Grunnleggende",
+    "network_ll_dp_description": "Kontroller hvilke TLV-er som skal sendes over Link Layer Discovery Protocol",
+    "network_ll_dp_disabled": "Funksjonshemmet",
```
"network_ll_dp_disabled": "Deaktivert",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "network_mac_address_title": "MAC-adresse",
+    "network_mdns_auto": "Bil",
+    "network_mdns_description": "Kontrollmodus for mDNS (multicast DNS)",
+    "network_mdns_disabled": "Funksjonshemmet",
```
"network_mdns_auto": "Auto",
    "network_mdns_description": "Kontrollmodus for mDNS (multicast DNS)",
    "network_mdns_disabled": "Deaktivert",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "updates_failed_get_device_version": "Klarte ikke å hente enhetsversjon: {error}",
+    "updating_leave_device_on": "Vennligst ikke slå av enheten din ...",
+    "usb": "USB",
+    "usb_config_custom": "Skikk",
```
"usb_config_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "usb_config_vendor_id_placeholder": "Skriv inn leverandør-ID",
+    "usb_device_classes_description": "USB-enhetsklasser i den sammensatte enheten",
+    "usb_device_classes_title": "Klasser",
+    "usb_device_custom": "Skikk",
```
"usb_device_custom": "Tilpasset",

adamshiervani commented on ui/localization/messages/nb.json:
```diff
@@ -0,0 +1,876 @@
+    "video_description": "Konfigurer skjerminnstillinger og EDID for optimal kompatibilitet",
+    "video_edid_acer_b246wl": "Acer B246WL, 1920x1200",
+    "video_edid_asus_pa248qv": "ASUS PA248QV, 1920x1200",
+    "video_edid_custom": "Skikk",
```
"video_edid_custom": "Tilpasset",

IDisposable commented:
I've got the documentation created and all the remaining translations I missed (I think?) done.

adamshiervani commented on DEVELOPMENT.md:
```diff
@@ -429,6 +441,79 @@ git diff --cached --diff-filter=d > ../internal/native/cgo/lvgl-minify.patch &&
+4. Save the _en.json_ file and execute `npm run i18n` to resort the language files, validate the translations, and create the m-functions
+5. Edit the _.tsx_ file and replace the string with the new m-function which will be the key-string you chose in snake-case. For example `This is a test` in _thing edit page_ turns into `m.thing_edit_this_is_a_test()`
+   - **Note** if the string has a replacement token, supply that to the m-function, for example `m.profile_your_name({ name: edit.value })`
+6. When all your strings are extracted, run `npm run i18n:machine-translate` to get a first-stab at the translations for the other supported languages.
```
Could you add that they should also use an LLM to confirm that the translation makes sense in the other languages? They're good enough for tasks like this now

IDisposable replied:
Done. (also updated the copy in this PR's description)

IDisposable commented:
Rebased to pull in the new sleep mode settings.

IDisposable commented on ui/src/routes/devices.$id.settings.video.tsx:
```diff
@@ -176,52 +170,52 @@ export default function SettingsVideoRoute() {
                   step="0.1"
                   value={videoSaturation}
-                  onChange={e => setVideoSaturation(parseFloat(e.target.value))}
+                  onChange={e => setVideoSaturation(Number.parseFloat(e.target.value))}
```
SonarQube thinks we should be explicit, leaving this as-is

IDisposable commented on ui/src/routes/devices.$id.settings.network.tsx:
```diff
@@ -47,11 +50,13 @@ const resolveOnRtcReady = () => {
   useEffect(() => {
-    setRemaining(dayjs(lifetime).fromNow());
+    // schedule immediate initial update
+    setInterval(() => setRemaining(dayjs(lifetime).fromNow()), 0);
```
Doing that results in a React warning about synchronously setting state. This eliminates that warning.

IDisposable commented on ui/src/routes/devices.$id.mount.tsx:
```diff
@@ -398,6 +397,12 @@ function UrlView({
 
   const urlRef = useRef<HTMLInputElement>(null);
 
+  useEffect(() => {
+    if (urlRef.current) {
+      setIsUrlValid(urlRef.current.validity.valid);
+    }
+  }, [url]);
```
This may be valid, but I don't really want to pull that into this PR

IDisposable commented on ui/tools/find_unused_messages.py:
```diff
@@ -0,0 +1,114 @@
+        payload = json.load(f)
+
+    keys = sorted(list(flatten(payload)))
+    keys.pop(0) if keys and keys[0] == "$schema" else None  # remove $schema if present
```
This is the standard idiom for Python (even if I find it ugly). Leaving it.

IDisposable commented on ui/src/routes/devices.$id.settings.network.tsx:
```diff
@@ -199,60 +199,84 @@ export default function SettingsNetworkRoute() {
+      });
+    }
+
+    if (dirty.ipv4_static?.gateway) {
```
Good catch! Fixed.

IDisposable commented on ui/src/routes/devices.$id.settings.network.tsx:
```diff
@@ -199,60 +199,84 @@ export default function SettingsNetworkRoute() {
+      changes.push({
+        label: m.network_ipv6_dns(),
+        from: initialSettingsRef.current?.ipv6_static?.dns.join(", ").toString() ?? "",
+        to: data.ipv4_static?.dns.join(", ").toString() ?? "",
```
Good catch. Fixed.

IDisposable committed:
Revamp the OTA and reboot processing

OTA supplies port-reboot action to handle the rebooting device.
Make sure we force-reload the page after redirection.
Move reboot logic into hw.go and make it set the willReboot message with parameters if provided.
Improve logging consistency.
